src/utils/PathUtils.ts

```diff
@@ -0,0 +1,141 @@
+import { PathPart } from '../fhirtypes';
+import { splitOnPathPeriods } from '../fhirtypes/common';
+import { CaretValueRule, Rule } from '../fshtypes/rules';
+
+/**
+ * Parses a FSH Path into a more easily usable form
+ * @param {string} fshPath - A syntactically valid path in FSH
+ * @returns {PathPart[]} an array of PathParts that is easier to work with
+ */
+export function parseFSHPath(fshPath: string): PathPart[] {
+  const pathParts: PathPart[] = [];
+  const seenSlices: string[] = [];
+  const indexRegex = new RegExp('^[0-9]$');
+  const splitPath = fshPath === '.' ? [fshPath] : splitOnPathPeriods(fshPath);
+  let depth = 0;
+  for (const pathPart of splitPath) {
+    const splitPathPart = pathPart.split('[');
+    if (splitPathPart.length === 1 || pathPart.endsWith('[x]')) {
+      // There are no brackets, or the brackets are for a choice, so just push on the name
+      pathParts.push({ base: pathPart, pathPosition: depth++ });
+    } else {
+      // We have brackets, let's  save the bracket info
+      let fhirPathBase = splitPathPart[0];
+      // Get the bracket elements and slice off the trailing ']'
+      let brackets = splitPathPart.slice(1).map(s => s.slice(0, -1));
+      // Get rid of any remaining [x] elements in the brackets
+      if (brackets[0] === 'x') {
+        fhirPathBase += '[x]';
+        brackets = brackets.slice(1);
+      }
+      brackets.forEach(bracket => {
+        if (!indexRegex.test(bracket) && !(bracket === '+' || bracket === '=')) {
+          seenSlices.push(bracket);
+        }
+      });
+      if (seenSlices.length > 0) {
+        pathParts.push({
+          base: fhirPathBase,
+          brackets: brackets,
+          slices: seenSlices,
+          pathPosition: depth++
+        });
+      } else {
+        pathParts.push({ base: fhirPathBase, brackets: brackets, pathPosition: depth++ });
+      }
+    }
+  }
+  return pathParts;
+}
+
+/**
+ * Assembles a PathPart array back to its original form
+ * @param {PathPart[]} pathParts - An array of pathParts
+ * @returns {string} path - The path corresponding to those pathParts
+ */
+export function assembleFSHPath(pathParts: PathPart[]): string {
+  let path = '';
+  pathParts.forEach((pathPart, i) => {
+    path += pathPart.base;
+    pathPart.brackets?.forEach(bracket => (path += `[${bracket}]`));
+    if (i < pathParts.length - 1) {
+      path += '.';
+    }
+  });
+  return path;
+}
+
+/**
+ *
+ * @param {PathPart} element - A single element in a rules path
+ * @param {Map<string, number} pathMap - A map containing an element's name as the key and that element's updated index as the value
+ */
+function convertSoftIndexes(element: PathPart, pathMap: Map<string, number>) {
+  // Must account for a pathPart's base name, it's position in the path, as well as any slices it's contained in.
+  const mapName = element.slices
+    ? `${element.base}${element.slices.join('|')}${element.pathPosition}`
+    : `${element.base}${element.pathPosition}`;
```

Comment:
Using the position in the path is clever, but unfortunately is still not unique enough.  For example, consider `ValueSet` which can have the following two possible paths:
```
compose.include[+].concept[+]
compose.exclude[+].concept[+]
```
The depth of `concept[+]` is `1` in both paths, but they reference unique arrays, so need to be tracked separately.  I think you're going to need to index on the resolved path prefix (e.g., `compose.include[0]` and `compose.exclude[0]`) instead of the depth.  If you did that, then the above could be something like:
```ts
const mapName = `${element.prefix ?? ''}.${element.base}|${(element.slices ?? []).join('|')}`
```

Reply:
This solution worked well, and I believe it's pretty air tight in regards to edge cases. Thank you for the suggestion!